File: src/python/oop/VendingMachine/vending/states/select_product_state.py
from ..order_context import OrderContext
from .order_state import OrderState


class SelectProductState(OrderState):
    def __init__(self, context: OrderContext):
        super().__init__(context)

    def process(self):
        if not self.context.rack_code:
            self.context.vm._notify("Please select product code")
            return

        try:
            rack = self.context.vm._inventory_manager.get_rack(self.context.rack_code)
            if not rack.is_empty:
                self.context.rack = rack
                self.context.vm._notify(f"Selected rack: {self.context.rack.number}")

                from .process_payment_state import ProcessPaymentState
                self._next_state(ProcessPaymentState(self.context), True)
                return

            self.context.vm._notify("The rack is empty")
        except ValueError as ex:
            self.context.vm._notify(str(ex))

        self.context.vm._notify("Please select product code")

    def state_info(self):
        self.context.vm._notify("Please select product code")

File: src/csharp/oop/VendingMachine/Vending/States/SelectProductState.cs
namespace VendingMachine.Vending.States;

public class SelectProductState(OrderContext context) : OrderStateBase(context)
{
    public override void Process()
    {
        if (Context.RackNumber == null)
        {
            Notify("Please select product code");
            return;
        }

        try
        {
            var rack = Context.VendingMachine.InventoryManager.GetRack(Context.RackNumber);
            if (rack.Count > 0)
            {
                Notify($"Selected rack: {Context.RackNumber}");

                Context.Rack = rack;
                NextState(new ProcessPaymentState(Context), true);
                return;
            }

            Notify($"Selected rack {Context.RackNumber} is empty");
        }
        catch (ArgumentException ex)
        {
            Notify(ex.Message);
        }

        Notify("Please select product code");
    }

    public override void StateInfo()
        => Notify("Please select product code");
}
